### PR TITLE
New medium track WR: 15.84min (-91s). Bigram hash embeddings, fused FP8 CE, unified optimizer

### PR DESCRIPTION
**Pull request by JOD-code**
I ported the highest-impact post-bulk-transfer improvements from the short track to the medium track, focused on the features with the largest measured impact:

* **Fused softcapped cross-entropy with FP8 pipeline** — FusedSoftcappedCrossEntropy from triton_kernels.py performs FP8 matmul + softcap + cross-entropy in one kernel. Combined with transposed lm_head layout (CastedLinearT) for faster gradient accumulation. Short track record 59: -0.9s.
* **Bigram hash embeddings** with sparse gradient communication (~80% sparsity reduction). Short track record 62: -5.6s.
* **Unified NorMuonAndAdam optimizer** with parameter banks, replacing 3 separate optimizers. Required as prerequisite for bigram's sharded_sparse comms. Short track record 61: -1.0s.
* Step count reduction 4740 → 4620 (enabled by loss budget from bigram)
* `coordinate_descent_tuning` enabled (~5s of total improvement; the rest came from the architectural changes)

I also tested Muon beta2 0.9 (short track record 78). It helps early training but washes out by step 3000 at this step count — kept at 0.95. See README for details.

Overlaps with PR #248 on FP8 lm_head (CastedLinearT). Remaining changes are orthogonal — #248's LR tuning and FP8 scale profiling could be adopted on top for further gains. Loss budget of 0.0094 below target leaves room for additional step reduction.

## Timing and Validation

```
import scipy.stats
import torch

losses = [2.9099, 2.9106, 2.9110, 2.9108, 2.9106]
times = [950.136, 950.425, 950.615, 949.573, 950.776]

print("p=%.4f" % scipy.stats.ttest_1samp(losses, 2.92, alternative="less").pvalue)
# p=0.0000

print("losses:", torch.std_mean(torch.tensor(losses)))
# losses: (tensor(0.0004), tensor(2.9106))

print("time:", torch.std_mean(torch.tensor(times)))
# time: (tensor(0.4735), tensor(950.3050))
```

timing prior record: 1041.277

---

### CODE REVIEW

**Comment by trianxy** [comment]
I see your PR suggests

> torch._inductor.config.coordinate_descent_tuning = True # allowed on medium track, ~30+ min extra compile time

I didn't notice this has been allowed on medium track. Has this been mentioned somewhere in the repo?

30min extra compile time is a problem for quick experimentation.

**Comment by JOD-code** [comment]
Yes, under the "Rules" heading in the main project's README it says:

> Note: torch._inductor.config.coordinate_descent_tuning is allowed for GPT-2 Medium track (a.k.a. 2.92 track).

I did all of my experimentation with it off. I only turned it on when I was doing the five runs in a row to get the p-value. It saved 5 seconds. 

I didn't actually time how long the compilation took. The estimate for the 30min is also taken from the main project's README:

> Not use any extra torch._inductor.config or torch.compile flags. (These can save a few seconds, but they can also make compilation take >30min. This rule was introduced after the 21st record.)

**Comment by trianxy** [comment]
Thank you @JOD-code , I had missed that note
